src/python/pants/backend/python/target_types_rules.py

```diff
@@ -141,23 +142,67 @@ class InjectPythonDistributionDependencies(InjectDependenciesRequest):
 
 @rule
 async def inject_python_distribution_dependencies(
-    request: InjectPythonDistributionDependencies,
+    request: InjectPythonDistributionDependencies, python_infer_subsystem: PythonInferSubsystem
 ) -> InjectedDependencies:
-    """Inject any `.with_binaries()` values, as it would be redundant to have to include in the
-    `dependencies` field."""
+    """Inject dependencies that we can infer from entry points in the distribution.
+
+    Inject module owners referred to from `entry_points`.
+
+    (`.with_binaries()` is deprecated) Inject any `.with_binaries()` values, as
+    it would be redundant to have to include in the `dependencies` field.
+    """
     original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
+
+    entry_points = original_tgt.target[PythonDistributionEntryPoints].value
+    entry_point_addresses = Addresses()
+    if entry_points and python_infer_subsystem.entry_points:
+        address = original_tgt.target.address
+        explicitly_provided_deps = await Get(
+            ExplicitlyProvidedDependencies, DependenciesRequest(original_tgt.target[Dependencies])
+        )
+        all_entry_points = [
+            (key, name, entry_point)
+            for key, section in entry_points.items()
+            for name, entry_point in section.items()
+        ]
+        all_owners = await MultiGet(
+            Get(PythonModuleOwners, PythonModule(entry_point.module))
+            for _, _, entry_point in all_entry_points
+        )
+        for (key, name, entry_point), owners in zip(all_entry_points, all_owners):
+            field_str = repr({key: {name: entry_point.spec}})
+            explicitly_provided_deps.maybe_warn_of_ambiguous_dependency_inference(
+                owners.ambiguous,
+                address,
+                import_reference="module",
+                context=(
+                    f"The distribution target {address} has the field "
+                    f"`entry_points={field_str}`"
+                ),
+            )
+            maybe_disambiguated = explicitly_provided_deps.disambiguated_via_ignores(
+                owners.ambiguous
+            )
+            unambiguous_owners = Addresses(
+                owners.unambiguous
+                or ((maybe_disambiguated,) if maybe_disambiguated else Addresses())
+            )
+            entry_point_addresses += unambiguous_owners  # type: ignore[assignment]
```

Comment:
I failed to satisfy mypy here...
```
src/python/pants/backend/python/target_types_rules.py: note: In function "inject_python_distribution_dependencies":
src/python/pants/backend/python/target_types_rules.py:189:37: error:
Incompatible types in assignment (expression has type "Tuple[Address, ...]",
variable has type "Addresses")  [assignment]
                entry_point_addresses = entry_point_addresses + unambiguou...
                                        ^
```

Reply:
I think it fails because `Addresses` should be immutable. Instead, you can use a list[Address] or a set[Address]. After you've finished the loop, convert it to an `Addresses` object. This will allow you to simplify the line above to avoid the awkward wrapping `Addresses` calls.